Workiva/over_react
# [2.0.0-dev - dart2] AF-233 Consume react-dart callable class fixes

> Depends on https://github.com/cleandart/react-dart/pull/142

### Problem(s):
1. In the Dart 2 SDK, the assignment of an object with a call method to a function type will now be treated as an implicit closurization of the `.call` method of the object.  This is not compatible with how `ReactComponentFactory` works.
2. The `MapBase` additions from #151are no longer the best way to satisfy the missing concrete `Map` class impls.
3. Misc. housekeeping items need to be addressed to get the `2.0.0-dev` branch ready to support further Dart 2 SDK testing / `2.0.0-dev.<n>` releases as the Dart team moves closer and closer to a production release of the Dart 2 SDK.

### Solution(s):

1. Consume the https://github.com/cleandart/react-dart/pull/142 callable class fixes
2. The `MapBase` additions from #151 were reverted and the new fields / getters / methods from that are present on the `Map` class in the Dart 2 SDK were instead implemented individually so we don't add a bunch of stuff from `dart:collection` that we don't actually need.
3. Address housekeeping items
    1. Update Dart SDK dependency range to `>=2.0.0-dev.0 <=2.0.0-dev.49.0`
    2. Address any lints / hints / errors found when the analysis server in Dart `w.0.0-dev.49.0` runs

### Testing Instructions:

1. Switch your Dart SDK to `2.0.0-dev.49.0`
3. Serve the examples using the DDC:
    ```bash
    pub serve --web-compiler=dartdevc
    ```
4. Load `http://localhost:8080` and play around with the example components.
5. Verify that the examples work as expected without producing run time errors


### Known Issues:

There are some issues present that, upon some investigation, appear to be unrelated to these changes:

1. There are numerous test failures centered around type casting.  I am unsure if this is a `test` package issue, or if we need to do away with some of our "shared" test suites.

---

__FYA:__ @greglittlefield-wf @clairesarsam-wf 

__FYI:__ @dustinlessard-wf @sebastianmalysa-wf @robbecker-wf
